docs/upgrade-notes/upgrade-execution-table-CUMULUS-3320.md

```diff
@@ -0,0 +1,284 @@
+
+#### Indexing Failure
+
+If the concurrent index query fails for any reason, you may have an `invalid` index - if this occurs, make sure to drop and/or recreate the index to avoid resources being used for the invalid index:
+
+```sql
+DROP INDEX CONCURRENTLY executions_collection_cumulus_id_index
+```
+
+The index operation can then be re-attempted.
+
+## Updating the `executions_parent_cumulus_id_foreign` constraint
+
+### *Notes*:
+
+- This update may require ingest downtime as updates to the table require some time.   See instructions below.
+- This update should be performed after adding the `executions_parent_cumulus_id_foreign` constraint
+
+### Reasoning
+
+This constraint as exists provides no action on deletion, this results in deletion of parent-child execution trees being both onerous and non-performant.   During work/evaluation for the CUMULUS-3320 feature implementation it was determined that adding `ON DELETE SET NULL` to this key makes wholesale/bulk deletes relatively performant (when combined with an index on the field) for large numbers of records while also being acceptable for 3rd party/smaller RESTful queries.
+
+### Procedure
+
+Users may opt to either:
+
+#### Utilize the normal cumulus deployment
+
+The cumulus module will run the migrations on the database when it's deployed.   For databases with smaller holdings and not under heavy load such that the sum of the migrations to be run will complete in < 15 minutes, the migration will run the following SQL query *inside a transaction*.   This query will write-lock the table, which may cause write failures due to timeouts if ingest is ongoing:
+
+```sql
+ALTER TABLE executions DROP CONSTRAINT IF EXISTS executions_parent_cumulus_id_foreign
+ALTER TABLE executions ADD CONSTRAINT executions_parent_cumulus_id_foreign FOREIGN KEY (parent_cumulus_id) REFERENCES executions(cumulus_id) ON DELETE SET NULL';
+```
+
+#### Manually update the constraint
+
+To manually update the constraint, do the following, in order:
+
+##### Modify the existing constraint without validation
+
+The following query will, in a transaction, remove the existing constraint and re-enable the updated constraint.   Adding `NOT VALID` to the creation will cause postgres not to validate values in existing rows, but will enforce them for new rows.    This *greatly* reduces downtime requirements as the constraint can be validated while the table is in use, and this transaction should be extremely quick:
+
+```sql
+BEGIN;
+ALTER TABLE executions DROP CONSTRAINT IF EXISTS executions_parent_cumulus_id_foreign;
+ALTER TABLE executions ADD CONSTRAINT executions_parent_cumulus_id_foreign FOREIGN KEY (parent_cumulus_id) REFERENCES executions(cumulus_id) ON DELETE SET NULL NOT VALID;
+END;
+```
+
+#### Re-validate the constraint
+
+This query can be run at any time following the above step and will not lock the full table for new row inserts/selects:
```

Comment:
Maybe a note about why we need to validate? Or is that too close to 'how to SQL'?

Reply:
Eh, it's mostly explained above.

Reply:
There's definitely an audience background concern here.   I think I fall on the side of YAGNI on this, but if you feel strongly I can try to include some links to postgres documents.